### PR TITLE
663/mk/open payments validate

### PR DESCRIPTION
**Pull request by mkurapov**
## Changes proposed in this pull request

- Adding Open API validation for response messages when making Open Payments API calls (using the sibling `open-api` package)
- Refactoring folder structure

## Context

The related issues are https://github.com/interledger/rafiki/issues/663, and https://github.com/interledger/rafiki/issues/662 and the parent roadmap item is https://github.com/interledger/roadmap/issues/20. 

## Checklist

- [x] Related issues linked using `fixes #number`
- [x] Tests added/updated
- [ ] Documentation added
- [x] Make sure that all checks pass

### CODE REVIEW

File: packages/open-payments/src/client/ilp-stream-connection.test.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
+/* eslint-disable @typescript-eslint/no-empty-function */
```
Comment (mkurapov) [inline packages/open-payments/src/client/ilp-stream-connection.test.ts:1]: not sure how important these tests in particular are, but I thought it was worth making sure we are adding the correct path for the response validator

File: packages/open-payments/src/types.ts [hunk 2]
```diff
@@ -1,4 +1,7 @@
-import { components } from './generated/types'
+import { components, paths as Paths } from './generated/types'
+
+export const getPath = <P extends keyof Paths>(path: P): string =>
```
Comment (mkurapov) [inline packages/open-payments/src/types.ts:3]: this function gives autocomplete functionality to all of the possible (generated) Open payments API paths we have

File: packages/open-payments/src/client/index.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  ilpStreamConnection: {
+    get(args: GetArgs): Promise<ILPStreamConnection>
+  }
+}
+
+export const createClient = async (
+  args?: CreateOpenPaymentClientArgs
+): Promise<OpenPaymentsClient> => {
+  const axios = createAxiosInstance(args)
+  const openApi = await createOpenAPI(config.OPEN_PAYMENTS_OPEN_API_URL)
```
Comment (mkurapov) [inline packages/open-payments/src/client/index.ts:26]: using the same version to validate the Open API as was used for generating the types in the first place

File: packages/open-payments/src/client/incoming-payment.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+export const getIncomingPayment = async (
+  axios: AxiosInstance,
+  openApi: OpenAPI,
+  args: GetArgs
+): Promise<IncomingPayment> => {
+  return get(
+    axios,
+    args,
+    openApi.createResponseValidator({
```
Comment (wilsonianb) [inline packages/open-payments/src/client/incoming-payment.ts:14]: The response validator library uses Ajv, which recommends instantiating once at initialization (vs once per request here)
https://github.com/kogosoftwarellc/open-api/blob/9b58193f17f69a3f6b018a17c79a8c938b7628a2/packages/openapi-response-validator/index.ts#L79-L85
https://ajv.js.org/guide/managing-schemas.html#compiling-during-initialization
(We should probably document this in our `openapi` library...)

Comment (mkurapov) [reply packages/open-payments/src/client/incoming-payment.ts]: Good call, updated to create the validators just once. Will also update the docs.

File: packages/openapi/README.md [hunk 5]
```diff
@@ -15,3 +17,49 @@ From the monorepo root directory:
 ```shell
 pnpm --filter openapi test
 ```
+
+## Usage
+
+First, instantiate an `OpenAPI` validator object with a reference to your OpenAPI spec:
+
+```ts
+const openApi = await createOpenAPI(OPEN_API_URL)
```
Comment (wilsonianb) [inline packages/openapi/README.md:26]: `createOpenAPI` accepts either a filepath or url of the OpenAPI spec 
https://github.com/APIDevTools/json-schema-ref-parser/blob/main/lib/index.d.ts#L35

File: packages/openapi/README.md [hunk 6]
```diff
@@ -15,3 +17,49 @@ From the monorepo root directory:
...
+})
+
+validateRequest(data) // true or false
+
+const validateResponse = openApi.createResponseValidator({
+  path: '/resource/{id}',
+  method: HttpMethod.GET
+})
+
+validateResponse(data) // true or false
```
Comment (wilsonianb) [inline packages/openapi/README.md:44]: `data` would be the response body

File: packages/openapi/README.md [hunk 7]
```diff
@@ -15,3 +17,49 @@ From the monorepo root directory:
...
+
+Then, validate requests and responses as such:
+
+```ts
+const validateRequest = openApi.createRequestValidator({
+  path: '/resource/{id}',
+  method: HttpMethod.GET
+})
+
+validateRequest(data) // true or false
```
Comment (wilsonianb) [inline packages/openapi/README.md:37]:
```suggestion
validateRequest(req) // true or false
```

https://github.com/kogosoftwarellc/open-api/blob/master/packages/openapi-request-validator/index.ts#L17
https://github.com/kogosoftwarellc/open-api/blob/master/packages/openapi-types/index.ts#L28-L33

File: packages/open-payments/src/client/requests.ts [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+      throw new Error(errorMessage)
+    }
+
+    return data
+  } catch (error) {
+    console.log('Error when making Open Payments GET request', {
+      errorMessage: error?.message,
+      url
+    })
```
Comment (wilsonianb) [inline packages/open-payments/src/client/requests.ts:44]: Should this either not log or should `open-payments` support a configurable logger like `pino`?

Comment (mkurapov) [reply packages/open-payments/src/client/requests.ts]: Yup, added pino support. wanted to do it as a follow up but it's not too big

File: packages/open-payments/src/client/index.ts [hunk 9]
```diff
@@ -0,0 +1,46 @@
+import { createOpenAPI, OpenAPI } from 'openapi'
+import createLogger, { Logger, LevelWithSilent as LogLevel } from 'pino'
```
Comment (wilsonianb) [inline packages/open-payments/src/client/index.ts:2]:
```suggestion
import createLogger, { Logger } from 'pino'
```

File: packages/open-payments/src/client/requests.ts [hunk 10]
```diff
@@ -0,0 +1,55 @@
+import axios, { AxiosError, AxiosInstance } from 'axios'
```
Comment (wilsonianb) [inline packages/open-payments/src/client/requests.ts:1]:
```suggestion
import axios, { AxiosInstance } from 'axios'
```

**Review by wilsonianb — COMMENTED** [review]
Might need to merge/rebase on `main` to resolve backend build failure

**Review by wilsonianb — APPROVED** [review]

**Pull request merged**